File: src/models/scanned_invoice.nim
import std/[times]
import orm/orm

type
  ScannedInvoice* = object of Model
    direction*: string        # PURCHASE, SALE, UNKNOWN
    status*: string           # PENDING, PROCESSED, REJECTED
    document_number*: string
    document_date*: DateTime
    due_date*: DateTime
    # Vendor info
    vendor_name*: string
    vendor_vat_number*: string
    vendor_address*: string
    # Customer info
    customer_name*: string
    customer_vat_number*: string
    customer_address*: string
    # Amounts
    subtotal*: float
    total_tax*: float
    invoice_total*: float
    # Validation
    validation_status*: string
    vies_validation_message*: string
    requires_manual_review*: bool
    manual_review_reason*: string
    # File
    file_name*: string
    file_path*: string
    # Suggested accounts
    counterparty_account_id*: int64
    vat_account_id*: int64
    expense_revenue_account_id*: int64
    # Relations
    company_id*: int64
    counterpart_id*: int64
    journal_entry_id*: int64
    created_by_id*: int64
    created_at*: DateTime
    updated_at*: DateTime

proc newScannedInvoice*(
  direction = "UNKNOWN",
  status = "PENDING",
  document_number = "",
  document_date = now(),
  vendor_name = "",
  vendor_vat_number = "",
  vendor_address = "",
  customer_name = "",
  customer_vat_number = "",
  customer_address = "",
  subtotal = 0.0,
  total_tax = 0.0,
  invoice_total = 0.0,
  validation_status = "PENDING",
  file_name = "",
  company_id: int64 = 0,
  created_by_id: int64 = 0
): ScannedInvoice =
  ScannedInvoice(
    id: 0,
    direction: direction,
    status: status,
    document_number: document_number,
    document_date: document_date,
    due_date: now(),
    vendor_name: vendor_name,
    vendor_vat_number: vendor_vat_number,
    vendor_address: vendor_address,
    customer_name: customer_name,
    customer_vat_number: customer_vat_number,
    customer_address: customer_address,
    subtotal: subtotal,
    total_tax: total_tax,
    invoice_total: invoice_total,
    validation_status: validation_status,
    vies_validation_message: "",
    requires_manual_review: false,
    manual_review_reason: "",
    file_name: file_name,
    file_path: "",
    counterparty_account_id: 0,
    vat_account_id: 0,
    expense_revenue_account_id: 0,
    company_id: company_id,
    counterpart_id: 0,
    journal_entry_id: 0,
    created_by_id: created_by_id,
    created_at: now(),
    updated_at: now()
  )

File: scanner_service_java/src/main/java/bg/baraba/scanner/model/entity/ScannedInvoice.java
package bg.baraba.scanner.model.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;

@Entity
@Table(name = "scanned_invoices")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ScannedInvoice {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "session_id")
    private Long sessionId;

    @Column(name = "session_file_id")
    private Long sessionFileId;

    @Column(name = "company_uid", nullable = false)
    private String companyUid;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private InvoiceDirection direction;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private ProcessingStatus status = ProcessingStatus.PENDING;

    // Vendor info
    @Column(name = "vendor_name")
    private String vendorName;

    @Column(name = "vendor_vat_number")
    private String vendorVatNumber;

    @Column(name = "vendor_address")
    private String vendorAddress;

    // Customer info
    @Column(name = "customer_name")
    private String customerName;

    @Column(name = "customer_vat_number")
    private String customerVatNumber;

    @Column(name = "customer_address")
    private String customerAddress;

    // Invoice details
    @Column(name = "invoice_number")
    private String invoiceNumber;

    @Column(name = "invoice_date")
    private LocalDate invoiceDate;

    @Column(name = "due_date")
    private LocalDate dueDate;

    // Amounts
    @Column(precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(name = "total_tax", precision = 15, scale = 2)
    private BigDecimal totalTax;

    @Column(name = "invoice_total", precision = 15, scale = 2)
    private BigDecimal invoiceTotal;

    // VIES validation
    @Column(name = "vies_status")
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private ViesStatus viesStatus = ViesStatus.PENDING;

    @Column(name = "vies_validation_message")
    private String viesValidationMessage;

    @Column(name = "vies_company_name")
    private String viesCompanyName;

    @Column(name = "vies_company_address")
    private String viesCompanyAddress;

    @Column(name = "vies_validated_at")
    private Instant viesValidatedAt;

    // Selected accounts
    @Column(name = "counterparty_account_id")
    private Long counterpartyAccountId;

    @Column(name = "vat_account_id")
    private Long vatAccountId;

    @Column(name = "expense_revenue_account_id")
    private Long expenseRevenueAccountId;

    // Review flags
    @Column(name = "requires_manual_review")
    @Builder.Default
    private Boolean requiresManualReview = false;

    @Column(name = "manual_review_reason")
    private String manualReviewReason;

    @Column
    private String notes;

    // Processing metadata
    @Column(precision = 5, scale = 4)
    private BigDecimal confidence;

    @Column(name = "original_file_name")
    private String originalFileName;

    @Column(name = "journal_entry_id")
    private Long journalEntryId;

    // Audit
    @Column(name = "created_by_id")
    private Long createdById;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private Instant updatedAt = Instant.now();

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }

    public enum InvoiceDirection {
        PURCHASE, SALE
    }

    public enum ProcessingStatus {
        PENDING, VALIDATED, REJECTED, PROCESSED
    }

    public enum ViesStatus {
        PENDING, VALID, INVALID, NOT_APPLICABLE, ERROR
    }
}
